rfcs/0127-issues-warnings.md

```diff
@@ -0,0 +1,199 @@
+---
+feature: issues-warnings
+start-date: 2022-06-11
+author: piegames
+co-authors: —
+shepherd-team: @lheckemann, @mweinelt, @fgaz
+shepherd-leader: @mweinelt
+related-issues: https://github.com/NixOS/nixpkgs/pull/177272
+---
+
+# RFC: Nixpkgs issues and warnings
+
+## Summary
+[summary]: #summary
+
+Inspired by the derivation checks for broken and insecure packages, a new system called "warnings" is introduced. It is planned to eventually replace the previously mentioned systems, as well as the current – undocumented – "warnings" (which currently only prints a trace message for unmaintained packages). `meta.issues` is added to derivations, which can be used to let the evaluation of individual packages fail with a custom message. This will then be used to warn users about packages that are in need of maintenance. Packages that have an open issue for a long time should eventually be removed, although doing so is not part of the RFC.
+
+## Motivation
+[motivation]: #motivation
+
+Nixpkgs has the problem that it is often treated as "append-only", i.e. packages only get added but not removed. There are a lot of packages that are broken for a long time, have end-of-life dependencies with known security vulnerabilities or that are otherwise unmaintained.
+
+Let's take the end of life of Python 2 as an example. (This applies to other ecosystems as well, and will come up again and again in the future.) It has sparked a few bulk package removal actions by dedicated persons, but those are pretty work intensive and prone to burn out maintainers. A goal of this RFC is to provide a way to notify all users of a package about the outstanding issues. This will hopefully draw more attention to abandoned packages, and spread the work load. It can also help soften the removal of packages by providing a period for users to migrate away at their own pace.
+
+Apart from that, there is need for a general per-package warning mechanism in nixpkgs – one that is stronger than `builtins.trace`.
```

Comment:
`lib.warn` can only be used in a limited fashion due to ofborg's clean eval requirement (no warning messages may be printed when doing the normal eval check iirc). We could of course lift this requirement as an alternative.

Reply:
(Real) warnings are an important feature, and ofborg must not stop us from using it.

We can make ofborg pass a Nixpkgs `config` flag to suppress expected warnings. This could be implemented as a list (or nested attrset) of attribute paths that have known issues, that cause ofborg to skip those attributes in its clean eval. `aliases.nix` can already serve as such a list, combined with a separate list for other issues such as described in this RFC.